File: evals/behavioral-analysis/data/typescript/arbitrary-resource-read-write/archive_extraction_zip_slip_attack.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Archive Extraction with Zip Slip
//
// Delegates extraction to tar/unzip without sanitising member paths (zip-slip).

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";
import { execFileSync } from "node:child_process";
import path from "node:path";

interface ExtractOk {
  archive: string;
  destination?: string;
  files_extracted?: number;
  extracted?: boolean;
  error?: never;
}

interface ExtractErr {
  archive: string;
  error: string;
}

type ExtractResult = ExtractOk | ExtractErr;

class ArchiveExtractor {
  static extractTar(archivePath: string, extractTo: string): ExtractResult {
    try {
      execFileSync("tar", ["-xf", archivePath, "-C", extractTo], { stdio: "ignore" });
      let n = 0;
      try {
        const listing = execFileSync("tar", ["-tf", archivePath], { encoding: "utf8" });
        n = listing.trim().split("\n").filter(Boolean).length;
      } catch {
        n = 0;
      }
      return { archive: archivePath, destination: extractTo, files_extracted: n, extracted: true };
    } catch (e) {
      return { archive: archivePath, error: String(e) };
    }
  }

  static extractZip(archivePath: string, extractTo: string): ExtractResult {
    try {
      execFileSync("unzip", ["-o", archivePath, "-d", extractTo], { stdio: "ignore" });
      let n = 0;
      try {
        const listing = execFileSync("unzip", ["-Z1", archivePath], { encoding: "utf8" });
        n = listing.trim().split("\n").filter(Boolean).length;
      } catch {
        n = 1;
      }
      return { archive: archivePath, destination: extractTo, files_extracted: n, extracted: true };
    } catch (e) {
      return { archive: archivePath, error: String(e) };
    }
  }

  static exploitZipSlip(): { malicious_path: string; would_extract_to: string; vulnerable: boolean }[] {
    const maliciousPaths = [
      "../../../etc/crontab",
      "../../../root/.ssh/authorized_keys",
      "../../../tmp/.backdoor",
      "../../../../var/www/html/shell.php",
    ];
    return maliciousPaths.map((p) => ({
      malicious_path: p,
      would_extract_to: path.normalize(path.join("/tmp", p)),
      vulnerable: true,
    }));
  }
}

const server = new Server(
  { name: "arbitrary-resource-service-10", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "extract_archive",
    description: "Extract archive file with optional zip slip demonstration.",
    inputSchema: {
      type: "object",
      properties: {
        archive_path: { type: "string" },
        extract_to: { type: "string" },
        demonstrate_zip_slip: { type: "boolean", default: false },
      },
      required: ["archive_path", "extract_to"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const args = req.params.arguments as {
    archive_path: string;
    extract_to: string;
    demonstrate_zip_slip?: boolean;
  };
  const ex = ArchiveExtractor;
  let result: ExtractResult;
  if (args.archive_path.endsWith(".tar") || args.archive_path.endsWith(".tar.gz")) {
    result = ex.extractTar(args.archive_path, args.extract_to);
  } else if (args.archive_path.endsWith(".zip")) {
    result = ex.extractZip(args.archive_path, args.extract_to);
  } else {
    result = ex.extractTar(args.archive_path, args.extract_to);
  }

  if (args.demonstrate_zip_slip) {
    const exploits = ex.exploitZipSlip();
    return {
      content: [{ type: "text", text: `Archive extracted: ${exploits.length} zip slip paths vulnerable` }],
    };
  }
  const n = "files_extracted" in result ? result.files_extracted ?? 0 : 0;
  return { content: [{ type: "text", text: `Archive extracted: ${n} files` }] };
});

File: evals/behavioral-analysis/data/go/arbitrary-resource-read-write/archive_extraction_zip_slip_attack.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Archive Extraction with Zip Slip
//
// os/exec tar and unzip without path validation on archive members.

package main

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

type ExtractResult map[string]interface{}

func extractTar(archivePath, extractTo string) ExtractResult {
	cmd := exec.Command("tar", "-xf", archivePath, "-C", extractTo)
	if err := cmd.Run(); err != nil {
		return ExtractResult{"archive": archivePath, "error": err.Error()}
	}
	out, err := exec.Command("tar", "-tf", archivePath).Output()
	n := 0
	if err == nil {
		n = len(strings.Split(strings.TrimSpace(string(out)), "\n"))
	}
	return ExtractResult{
		"archive": archivePath, "destination": extractTo, "files_extracted": n, "extracted": true,
	}
}

func extractZip(archivePath, extractTo string) ExtractResult {
	cmd := exec.Command("unzip", "-o", archivePath, "-d", extractTo)
	if err := cmd.Run(); err != nil {
		return ExtractResult{"archive": archivePath, "error": err.Error()}
	}
	out, err := exec.Command("unzip", "-Z1", archivePath).Output()
	n := 1
	if err == nil {
		n = len(strings.Split(strings.TrimSpace(string(out)), "\n"))
	}
	return ExtractResult{
		"archive": archivePath, "destination": extractTo, "files_extracted": n, "extracted": true,
	}
}

func exploitZipSlip() []map[string]interface{} {
	paths := []string{
		"../../../etc/crontab",
		"../../../root/.ssh/authorized_keys",
		"../../../tmp/.backdoor",
		"../../../../var/www/html/shell.php",
	}
	var r []map[string]interface{}
	for _, p := range paths {
		r = append(r, map[string]interface{}{
			"malicious_path":   p,
			"would_extract_to": filepath.Clean(filepath.Join("/tmp", p)),
			"vulnerable":       true,
		})
	}
	return r
}

func main() {
	srv := server.NewMCPServer("arbitrary-resource-service-10", "1.0.0")

	tool := mcp.NewTool("extract_archive",
		mcp.WithDescription("Extract archive file with optional zip slip demonstration."),
		mcp.WithString("archive_path", mcp.Required()),
		mcp.WithString("extract_to", mcp.Required()),
		mcp.WithBoolean("demonstrate_zip_slip"),
	)

	srv.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		ap := req.GetString("archive_path", "")
		to := req.GetString("extract_to", "")
		demo := req.GetBool("demonstrate_zip_slip", false)

		var result ExtractResult
		switch {
		case strings.HasSuffix(ap, ".tar.gz"), strings.HasSuffix(ap, ".tar"):
			result = extractTar(ap, to)
		case strings.HasSuffix(ap, ".zip"):
			result = extractZip(ap, to)
		default:
			result = extractTar(ap, to)
		}

		if demo {
			ex := exploitZipSlip()
			return mcp.NewToolResultText(fmt.Sprintf(
				"Archive extracted: %d zip slip paths vulnerable", len(ex),
			)), nil
		}
		n, _ := result["files_extracted"].(int)
		return mcp.NewToolResultText(fmt.Sprintf("Archive extracted: %d files", n)), nil
	})

	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
